File: packages/sdk-python/src/melaya/hitl.py
"""HITL (Human-in-the-Loop) API — list, approve, and reject pending tool-call approvals.

Maps to ``/api/v1/private/hitl/*``.

Example
-------
>>> from melaya import Melaya
>>> m = Melaya(api_key="mk_...")
>>> pending = m.hitl.pending()
>>> for req in pending:
...     m.hitl.approve(req["requestId"], comment="Looks good")
>>> m.hitl.bulk_decide(request_ids=["r1", "r2"], decision="approved")
"""
from __future__ import annotations

from typing import Any, Dict, List, Optional

from .platform_types import HitlDecision, JsonDict


class HitlAPI:
    def __init__(self, request: Any) -> None:
        self._request = request

    def pending(self) -> List[JsonDict]:
        """List all pending HITL tool-call approvals for the authenticated user."""
        return self._request("GET", "/api/v1/private/hitl/approvals/pending")

    def history(self, *, limit: Optional[int] = None, offset: Optional[int] = None) -> List[JsonDict]:
        """List historical HITL approval decisions."""
        params: Dict[str, Any] = {}
        if limit is not None:
            params["limit"] = limit
        if offset is not None:
            params["offset"] = offset
        return self._request("GET", "/api/v1/private/hitl/approvals/history", params=params)

    def approve(self, request_id: str, *, comment: Optional[str] = None) -> JsonDict:
        """Approve a pending HITL tool-call approval."""
        body: Dict[str, Any] = {}
        if comment is not None:
            body["comment"] = comment
        return self._request("POST", f"/api/v1/private/hitl/approvals/{request_id}/approve",
                             json=body if body else None)

    def reject(self, request_id: str, *, comment: Optional[str] = None) -> JsonDict:
        """Reject a pending HITL tool-call approval."""
        body: Dict[str, Any] = {}
        if comment is not None:
            body["comment"] = comment
        return self._request("POST", f"/api/v1/private/hitl/approvals/{request_id}/reject",
                             json=body if body else None)

    def bulk_decide(
        self,
        *,
        request_ids: List[str],
        decision: HitlDecision,
        comment: Optional[str] = None,
    ) -> JsonDict:
        """Bulk approve or reject multiple pending tool calls in one call."""
        body: Dict[str, Any] = {"requestIds": request_ids, "decision": decision}
        if comment is not None:
            body["comment"] = comment
        return self._request("POST", "/api/v1/private/hitl/approvals/bulk", json=body)

    # ── Run-level inspection ─────────────────────────────────────────────────────

    def run_tool_stats(self, run_id: str) -> JsonDict:
        """Get tool-call statistics for a specific pipeline run."""
        return self._request("GET", f"/api/v1/private/hitl/runs/{run_id}/tool-stats")

    def run_messages(
        self,
        run_id: str,
        *,
        limit: Optional[int] = None,
        cursor: Optional[str] = None,
    ) -> List[JsonDict]:
        """Get paginated messages for a pipeline run."""
        params: Dict[str, Any] = {}
        if limit is not None:
            params["limit"] = limit
        if cursor is not None:
            params["cursor"] = cursor
        return self._request("GET", f"/api/v1/private/hitl/runs/{run_id}/messages", params=params)

    def run_tool_stats_by_agent(self, run_id: str) -> JsonDict:
        """Get tool-call stats broken down by agent for a run."""
        return self._request("GET", f"/api/v1/private/hitl/runs/{run_id}/tool-stats/by-agent")

    def run_tool_calls(self, run_id: str) -> List[JsonDict]:
        """Get all tool calls for a pipeline run."""
        return self._request("GET", f"/api/v1/private/hitl/runs/{run_id}/tool-calls")

File: packages/sdk-rust/src/hitl.rs
use serde_json::{json, Value};
use std::collections::HashMap;

use crate::client::HttpClient;
use crate::error::Result;

/// HITL (Human-in-the-Loop) API — list, approve, and reject pending tool-call approvals.
#[derive(Clone)]
pub struct HitlAPI {
    http: HttpClient,
}

impl HitlAPI {
    pub(crate) fn new(http: HttpClient) -> Self {
        Self { http }
    }

    /// List all pending HITL tool-call approvals for the authenticated user.
    pub async fn pending(&self) -> Result<Value> {
        let q = HashMap::new();
        self.http
            .get("/api/v1/private/hitl/approvals/pending", &q)
            .await
    }

    /// List historical (decided) HITL approval records.
    pub async fn history(&self, limit: Option<u32>, offset: Option<u32>) -> Result<Value> {
        let mut q: HashMap<&str, Option<String>> = HashMap::new();
        q.insert("limit", limit.map(|v| v.to_string()));
        q.insert("offset", offset.map(|v| v.to_string()));
        self.http
            .get("/api/v1/private/hitl/approvals/history", &q)
            .await
    }

    /// Approve a pending HITL tool-call approval.
    pub async fn approve(&self, request_id: &str, comment: Option<&str>) -> Result<Value> {
        let mut body = json!({});
        if let Some(c) = comment {
            body["comment"] = json!(c);
        }
        self.http
            .post(
                &format!("/api/v1/private/hitl/approvals/{request_id}/approve"),
                &body,
            )
            .await
    }

    /// Reject a pending HITL tool-call approval.
    pub async fn reject(&self, request_id: &str, comment: Option<&str>) -> Result<Value> {
        let mut body = json!({});
        if let Some(c) = comment {
            body["comment"] = json!(c);
        }
        self.http
            .post(
                &format!("/api/v1/private/hitl/approvals/{request_id}/reject"),
                &body,
            )
            .await
    }

    /// Bulk approve or reject multiple pending approvals in one call.
    ///
    /// `decision` must be `"approved"` or `"rejected"`.
    pub async fn bulk_decide(
        &self,
        request_ids: &[&str],
        decision: &str,
        comment: Option<&str>,
    ) -> Result<Value> {
        let mut body = json!({
            "requestIds": request_ids,
            "decision": decision,
        });
        if let Some(c) = comment {
            body["comment"] = json!(c);
        }
        self.http
            .post("/api/v1/private/hitl/approvals/bulk", &body)
            .await
    }

    // ── Run-level inspection ─────────────────────────────────────────────────

    /// Get tool-call statistics for a specific pipeline run.
    pub async fn run_tool_stats(&self, run_id: &str) -> Result<Value> {
        let q = HashMap::new();
        self.http
            .get(
                &format!("/api/v1/private/hitl/runs/{run_id}/tool-stats"),
                &q,
            )
            .await
    }

    /// Get paginated messages for a pipeline run.
    pub async fn run_messages(
        &self,
        run_id: &str,
        limit: Option<u32>,
        cursor: Option<&str>,
    ) -> Result<Value> {
        let mut q: HashMap<&str, Option<String>> = HashMap::new();
        q.insert("limit", limit.map(|v| v.to_string()));
        q.insert("cursor", cursor.map(str::to_owned));
        self.http
            .get(&format!("/api/v1/private/hitl/runs/{run_id}/messages"), &q)
            .await
    }

    /// Get tool-call stats broken down by agent for a run.
    pub async fn run_tool_stats_by_agent(&self, run_id: &str) -> Result<Value> {
        let q = HashMap::new();
        self.http
            .get(
                &format!("/api/v1/private/hitl/runs/{run_id}/tool-stats/by-agent"),
                &q,
            )
            .await
    }

    /// Get all tool calls for a pipeline run.
    pub async fn run_tool_calls(&self, run_id: &str) -> Result<Value> {
        let q = HashMap::new();
        self.http
            .get(
                &format!("/api/v1/private/hitl/runs/{run_id}/tool-calls"),
                &q,
            )
            .await
    }
}
